Core, Spark: Remove dangling deletes as part of RewriteDataFilesAction

dramaticlly opened the pull request:
Goal: Attempt to clean up the dangling deletes as part of Spark RewriteDataFilesAction, it can be controlled by feature flag
`remove-dangling-deletes` and it's by default turned on. Most of the code come from #6581 and reason on why we need it: The problem and design doc is here: https://docs.google.com/document/d/11d-cIUR_89kRsMmWnEoxXGZCvp7L4TUmPJqUC60zB5M/edit#

Changes
- DeleteFiles to remove a given deleteFile 
- RewriteDataFilesResult now provide count on number of dangling files removed 
- withReusableDS() function moved from RewriteManifestsSparkAction to base so it can be reused in RewriteDataFilesAction.

TODO, figure out predicate push down for entries metadata table. 
-  rewriteDataFiles today support a custom expression filter which are targeted against base table schema
- entries metadata table used for identifying the dangling deletes will have a different schema than base table, the partition filter can be translated but no sure about filter applied on other non-partitioned columns.

dramaticlly commented:
@szehon-ho can I ask for your eyes first?

ajantha-bhat commented on api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java:
```diff
@@ -99,6 +99,20 @@ public interface RewriteDataFiles
 
+  /**
+   * Removes dangling delete files from current snapshot after compaction. A delete file is
+   * considered dangling if it does not apply to any non-expired data file.
```
Rewrite operation never used to delete any files. So, deleting some files from rewrite might confuse users? 

Is it possible to clean these files from expire snapshots (or as orphan files). 
Maybe we can just update the metadata by making these unreferenced and the actual delete can happen from GC functions.

manuzhang replied:
"delete files" here means delete files from current(live) snapshot. Rewrite data files can already delete equality delete files from current snapshot. It doesn't remove files from filesystem.

manuzhang replied:
"delete files" here means delete files from current(live) snapshot. Rewrite data files can already delete equality delete files from current snapshot. It doesn't remove files from filesystem.

manuzhang replied:
>  A delete file is considered dangling if it does not apply to any non-expired data file.
I think the most accurate wording is `if it does not apply to data files of current snapshot`.

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteDataFilesAction.java:
```diff
@@ -103,17 +108,21 @@
+import org.apache.spark.sql.Encoders;
 import org.apache.spark.sql.Row;
 import org.apache.spark.sql.internal.SQLConf;
+import org.junit.jupiter.api.Assertions;
```
this should be AssertJ assertions (which is already imported)

szehon-ho commented on api/src/main/java/org/apache/iceberg/RemoveDanglingDeletesMode.java:
```diff
@@ -0,0 +1,64 @@
+ * <p>
+ *
+ * <ul>
+ *   <li>If remove-dangling-deletes=metadata, then dangling delete files will be pruned from iceberg
```
maybe can remove 'remove-dangling-deletes=' , its a bit repetitive?

ie,
```
metadata: dangling delete files will be pruned from...
```

szehon-ho commented on api/src/main/java/org/apache/iceberg/RemoveDanglingDeletesMode.java:
```diff
@@ -0,0 +1,64 @@
+ *
+ * <p>
+ */
+public enum RemoveDanglingDeletesMode {
```
@aokolnychyi the thought here, is we will have further modes:  STATS (for partition file stats), FULL (run a whole rewritePositionDeletes job)

Default is NONE because it does create one more snapshot and will might break people who are depending on listing the snapshots for information.

szehon-ho commented on api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java:
```diff
@@ -99,6 +100,24 @@ public interface RewriteDataFiles
+   * <p>
+   *
+   * <ul>
+   *   <li>If remove-dangling-deletes=metadata, then dangling delete files will be pruned from
```
Same small comment from above javadoc (remove 'remove-dangling-deletes=' as its too repetitive)

ie, 
```
metadata: dangling delete files will be pruned from...
```

szehon-ho commented on api/src/main/java/org/apache/iceberg/RemoveDanglingDeletesMode.java:
```diff
@@ -0,0 +1,64 @@
+    // Replace the hyphen in order name with underscore to map to the enum value. For example:
+    // rewrite-position to REWRITE_POSITION
+    try {
+      return RemoveDanglingDeletesMode.valueOf(
```
this is not needed anymore right?  (The enum values have no hyphen )

dramaticlly replied:
yes you are right. Initially i though add `rewrite-position` so that we can trigger rewrite position deletes spark action but removed later. I guess I can remove it now and add it later

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -323,10 +350,117 @@ private Result doExecute(
+    return ImmutableRewriteDataFiles.Result.builder().rewriteResults(rewriteResults);
+  }
+
+  private List<DeleteFile> removeDanglingDeletes() {
```
Do you think we can move this to another , package protected class like RemoveDanglingDeleteSparkAction?  (for better code reading)

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -507,4 +645,54 @@ public int totalGroupCount() {
     }
   }
+
+  private static class MakeDeleteFile implements MapFunction<Row, DeleteFile> {
```
@zinking brought up a good point in #6581 that this may not be necessary if we can remove a delete file by path.   It can probably reduce memory pressure and allow us not to project unnecessary fields , for example  fileSize, here.

@dramaticlly pointed out that the call seems not to work for delete files.  ref: https://github.com/apache/iceberg/blob/apache-iceberg-1.4.3/core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java#L205

@aokolnychyi any thoughts here?

aokolnychyi replied:
Replied [here]. I think we use use `SparkDeleteFile` to wrap Spark rows as delete files.

szehon-ho reviewed (commented):
Hi @dramaticlly thanks for doing this.  Can you put me as a co-author as most of the code is from #6581 ?  

And as that is the case, it would be nice if @aokolnychyi takes a look as well

zinking commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -507,4 +645,54 @@ public int totalGroupCount() {
+      long fileSize = row.getAs(3);
+      long recordCount = row.getAs(4);
+
+      FileMetadata.Builder builder = FileMetadata.deleteFileBuilder(specsById.get(specId));
```
I probably didn't explain it properly. 
the `partition` variable and the `specsById.get(specId)` doesn't match under schema (partition) evolution. 

my fix is something like this 
```
      PartitionSpec partitionSpec = specsById.get(specId);
      List<String> partitionFields =
          partitionSpec.fields().stream().map(PartitionField::name).collect(Collectors.toList());
      PartitionData specPartData = partition.project(partitionFields);
      FileMetadata.Builder builder = FileMetadata.deleteFileBuilder(partitionSpec);
      ```

dramaticlly replied:
I see what you mean, because of the fact that join condition was based on specId and partition data to match, this will not remove dangling deletes after partition evolution where rewrite data files rewrite data files into new partition spec. Quick example illustrated below where even though we have equality delete with sequence number of 3 smaller than data files min sequence number of 5, due to partition evolution, the dangling equality delete will not be evaluated in this case. 

Instead of looking into project partition data into new struct, maybe we can see for how to remove delete file based on file path. 

```
equality delete entries()
+---------+-------+--------------------+-------+------------------+------------+---------------+
|partition|spec_id|           file_path|content|file_size_in_bytes|record_count|sequence_number|
+---------+-------+--------------------+-------+------------------+------------+---------------+
|{2, NULL}|      0|file:/var/folders...|      2|               465|           1|              4|
|{1, NULL}|      0|file:/var/folders...|      2|               465|           1|              3|
+---------+-------+--------------------+-------+------------------+------------+---------------+
```

```
minSeqNumberPerDataFilesPartition
+---------+-------+------------------------+
|partition|spec_id|min_data_sequence_number|
+---------+-------+------------------------+
|{1, CCCC}|      1|                       5|
|{1, BBBB}|      1|                       5|
|{2, DDDD}|      1|                       5|
+---------+-------+------------------------+
```

aokolnychyi replied:
Deleting based on path is not a good idea as Iceberg won't be able to prune manifests using partition info. The action for rewriting manifests already handles this, we can use a similar approach.

aokolnychyi commented:
I should have time to take a look this week.

aokolnychyi commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -278,20 +277,6 @@ private Dataset<Row> repartitionAndSort(Dataset<Row> df, Column col, int numPart
     return df.repartitionByRange(numPartitions, col).sortWithinPartitions(col);
   }
 
-  private <T, U> U withReusableDS(Dataset<T> ds, Function<Dataset<T>, U> func) {
```
I don't think this should be moved and used for this work.

aokolnychyi reviewed (commented):
Okay, I did one pass and here are my high-level notes:

- We should use `RewriteFiles` instead of `DeleteFiles`, changes in `DeleteFiles` should be reverted.
- I don't see a need for the enum to control the cleanup mode.
- I'd consider having a separate action but I can be convinced otherwise.
- I'd consider the following algorithm:
    - Extend `data_files` and `delete_files` metadata tables to include data sequence numbers, if needed. I don't remember if we already populate them. This should be trivial as each `DeleteFile` object already has this info.
    - Query `data_files`, aggregate, compute min data sequence number per partition. Don't cache the computed result, just keep a reference to it.
    - Query `delete_files`, potentially projecting only strictly required columns.
    - Join the summary with `delete_files` on the spec ID and partition. Find delete files that can be discarded in one go by having a predicate that accounts for the delete type (position vs equality).
    - Collect the result to the driver and use `SparkDeleteFile` to wrap Spark rows as valid delete files. See the action for rewriting manifests for an example.

zinking commented on api/src/main/java/org/apache/iceberg/DeleteFiles.java:
```diff
@@ -55,6 +55,17 @@ default DeleteFiles deleteFile(DataFile file) {
+   * @param file a DeleteFile to remove from the table
+   * @return this for method chaining
+   */
+  default DeleteFiles deleteFile(DeleteFile file) {
```
make sense + 1.

zinking commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -323,10 +350,117 @@ private Result doExecute(
+            .cache();
+
+    // Dangling position delete files
+    removedDeleteFiles.addAll(danglingPositionDeletes(entries, minDataSeqNumberPerPartition));
```
it's slightly different join condition where for eq delete it is <= and for pos it is <

aokolnychyi replied:
Is it the join or the filter condition? I think the join is on the partition and spec ID in both cases, and we can come up with a filter can check the delete type and use `<=` or `<`.

zinking replied:
> Is it the join or the filter condition? I think the join is on the partition and spec ID in both cases, and we can come up with a filter can check the delete type and use `<=` or `<`, instead of doing two scans.

right, that's correct. sounds correct thing todo.

nastra commented on api/src/main/java/org/apache/iceberg/RemoveDanglingDeletesMode.java:
```diff
@@ -0,0 +1,64 @@
+    // rewrite-position to REWRITE_POSITION
+    try {
+      return RemoveDanglingDeletesMode.valueOf(
+          modeName.replaceFirst("-", "_").toUpperCase(Locale.ENGLISH));
```
```suggestion
          modeName.toUpperCase(Locale.ENGLISH));
```

nastra commented on api/src/main/java/org/apache/iceberg/RemoveDanglingDeletesMode.java:
```diff
@@ -0,0 +1,64 @@
+
+  public static RemoveDanglingDeletesMode fromName(String modeName) {
+    String errorMessage = "Invalid remove dangling deletes mode name: %s";
+    Preconditions.checkArgument(modeName != null, String.format(errorMessage, "null"));
```
```suggestion
    Preconditions.checkArgument(modeName != null, errorMessage, "null");
```

nastra commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -172,11 +192,18 @@ public RewriteDataFiles.Result execute() {
 
     Stream<RewriteFileGroup> groupStream = toGroupStream(ctx, fileGroupsByPartition);
 
+    ImmutableRewriteDataFiles.Result.Builder result;
```
```suggestion
    ImmutableRewriteDataFiles.Result.Builder builder;
```

dramaticlly commented:
> Okay, I did one pass and here are my high-level notes:
> 
> * We should use `RewriteFiles` instead of `DeleteFiles`, changes in `DeleteFiles` should be reverted.
> * I don't see a need for the enum to control the cleanup mode.
> * I'd consider having a separate action but I can be convinced otherwise. Especially, given that we may account for partition stats in the future.
> * I'd consider the following algorithm:
>   
>   * Extend `data_files` and `delete_files` metadata tables to include data sequence numbers, if needed. I don't remember if we already populate them. This should be trivial as each `DeleteFile` object already has this info.
>   * Query `data_files`, aggregate, compute min data sequence number per partition. Don't cache the computed result, just keep a reference to it.
>   * Query `delete_files`, potentially projecting only strictly required columns.
>   * Join the summary with `delete_files` on the spec ID and partition. Find delete files that can be discarded in one go by having a predicate that accounts for the delete type (position vs equality).
>   * Collect the result to the driver and use `SparkDeleteFile` to wrap Spark rows as valid delete files. See the action for rewriting manifests for an example.

Based on Anton's feedback, I will try divide the changes into 2 PRs where first PR (#9813) to support data sequence number in data and delete files table. Once merged, I will update to scan data_files first to aggregate per spec/partition min data sequence number, then compare against the delete_files. With left join, we can identify dangling deletes and remove them in one pass. SparkDeleteFile will be used to convert from spark row to POJO to be used for pruning, in consideration of partition evolution. Finally, dangling delete will be removed by reconstruct instead of by file path, to benefit manifest pruning when iceberg table was scanned.

zinking commented:
"Finally, dangling delete will be removed by reconstruction instead of by file path, to benefit manifest pruning when iceberg table was scanned." 
I guess only partitionData and path is needed, others all not used.

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRemoveDanglingDeleteAction.java:
```diff
@@ -0,0 +1,438 @@
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.Encoders;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
```
please use Assertions from AssertJ

dramaticlly commented:
With the merge of #10203 , I refactored the algorithm a bit to scan entries table for getting minSequenceNumberPerPartitionAndSpec and for getting delete files table for data sequence number instead of rely on data sequence number as virtual columns. I also identified and fixed the problem in partition evolution tests so that now it's all handled correctly. Would you like to take another look? @szehon-ho @aokolnychyi

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -82,8 +83,9 @@ public class RewriteDataFilesSparkAction
           USE_STARTING_SEQUENCE_NUMBER,
-          REWRITE_JOB_ORDER,
-          OUTPUT_SPEC_ID);
+          OUTPUT_SPEC_ID,
```
question, why change the order?

szehon-ho replied:
Nit: still show up as changing order in this version

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -175,11 +178,21 @@ public RewriteDataFiles.Result execute() {
 
     Stream<RewriteFileGroup> groupStream = toGroupStream(ctx, fileGroupsByPartition);
 
+    ImmutableRewriteDataFiles.Result.Builder resultBuilder;
```
can we import  Result ?

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,175 @@
+      commit(rewriteFiles);
+    }
+
+    return new RemoveDanglingDeleteFilesActionResult(danglingDeletes);
```
Would it make sense to have just a count here?  

One reason being , if the user keeps a handle of the Result for some reason, the list wont be in the memory (it can be GC'ed after this method)?

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,175 @@
+                    .col("data_file.partition")
+                    .equalTo(minSequenceNumberByPartition.col("grouped_partition")));
+
+    Column filterCondition =
```
filterCondition => danglingDelete?

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,175 @@
+    return danglingDeletes.collectAsList().stream()
+        .map(
+            row ->
+                deleteFileWrapper(danglingDeletes.schema(), row.getInt(row.fieldIndex("spec_id")))
```
could we move the column index specific logic inside the deleteFileWrapper method?

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,175 @@
+
+  private SparkDeleteFile deleteFileWrapper(StructType sparkFileType, int specId) {
+    Types.StructType combinedFileType = DataFile.getType(Partitioning.partitionType(table));
+    // deleteFile need to use the same spec for which manifest was written in order for deletion
```
not 100% sure I get this comment, what does it mean 'which manifest was written'.  Can this be more succint/clear?

Is it something like `Set correct spec id`

szehon-ho reviewed (commented):
Looks mostly good to me, did a review round on the code portion.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+  private final Table table;
+
+  protected RemoveDanglingDeletesSparkAction(SparkSession spark, Table table) {
+    super(spark.cloneSession());
```
we may not need to cloneSession, I think?  It was done in RewriteDataFiles because it disabled AQE, but I think we dont do it here?

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+          .build();
+    }
+
+    String desc = String.format("Removing dangling delete in %s", table.name());
```
Nit: delete => delete files?

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+  /**
+   * Dangling delete files can be identified with following steps
+   *
+   * <p>1.Group data files by partition keys and find the minimum data sequence number in each
```
Is there a way to just use html numbered list here?

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+   *
+   * <p>2. Left outer join delete files with partition-grouped data files on partition keys.
+   *
+   * <p>3. Filter results to find dangling delete files by comparing delete file sequence_number to
```
Sorry for making another suggestion, how about:

``` Compare each delete file's sequence number to its partition's minimum data sequence number to find dangling deletes.```

Reading 'filter results to find ...' makes you keep too much in mind before getting to the point.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+  private List<DeleteFile> findDanglingDeletes() {
+    Dataset<Row> minSequenceNumberByPartition =
+        loadMetadataTable(table, MetadataTableType.ENTRIES)
+            // find live entry for data files
```
Nit: `find live data files`?   Again its a bit easier to me to get quickly

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,175 @@
+
+    Dataset<Row> deleteEntries =
+        loadMetadataTable(table, MetadataTableType.ENTRIES)
+            .filter(" data_file.content != 0 AND status < 2");
```
oops i see, nit: can you remove space in `" data`? 

And also comment => `find live delete files` for the same reason as above

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,175 @@
+
+    Column filterCondition =
+        col("min_data_sequence_number")
+            .isNull()
```
Thanks, can you also put that in the comment as well?  How about:

```delete fies without any data files in partition ```

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+                col("data_file.content")
+                    .equalTo("1")
+                    .and(col("sequence_number").$less(col("min_data_sequence_number"))))
+            // dangling equality delete files
```
same, these may be removed by first condition:

```equality delete files without any applicable data files in the partition```

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+        col("min_data_sequence_number")
+            // when all data files are rewritten with new partition spec
+            .isNull()
+            // dangling position delete files
```
hm, to clarify from the previous condition (also dangling), how about

```positional delete files without any applicable data files in partition```

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+            .or(
+                col("data_file.content")
+                    .equalTo("2")
+                    .and(col("sequence_number").$less$eq(col("min_data_sequence_number"))));
```
Actually, just read the spec that equality delete in unpartitioned spec is GLOBAL delete.  Will we remove that one by mistake, and can we add a test for that?

dramaticlly replied:
I think we checked for if table is partitioned at very [early](https://github.com/apache/iceberg/pull/9724/files#diff-6bf6c20da3c848d75bbba6bfb104d8399dde5197cadc580da25de15c867df60bR77-R82) and return early as ManifestFilterManager will handle such for unpartitioned table. Do you think it would be sufficient or do you want a partition evolution to move partitioned table into unpartitioned one with a dangling equality delete?

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+            .join(minSequenceNumberByPartition, joinOnPartition, "left")
+            .filter(filterOnDanglingDeletes)
+            .select("data_file.*");
+    return danglingDeletes.collectAsList().stream()
```
curious, is this something we can do on executors:

```
 deleteEntries
            .join(minSequenceNumberByPartition, joinOnPartition, "left")
            .filter(filterOnDanglingDeletes)
            .select("data_file.*")
            .map(row -> deleteFileWrapper(danglingDeletes.schema(), row))
```
?

dramaticlly replied:
oh I think we actually need to collect to driver and apply conversion, the reason is that the [`SparkDeleteFile`](https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkDeleteFile.java#L25)  is not serializable and @aokolnychyi told me to wrap after collect to driver.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -175,11 +179,21 @@ public RewriteDataFiles.Result execute() {
 
     Stream<RewriteFileGroup> groupStream = toGroupStream(ctx, fileGroupsByPartition);
 
+    Builder resultBuilder;
```
Nit: how about a method so we can still assign resultBuilder in one line?

``` Builder resultBuilder = doExecute(ctx, groupStream, ...)```

I guess we will have to rename doExecute .  doBatchExecute?

dramaticlly replied:
sure, I can rename and inline this

szehon-ho reviewed (commented):
Thanks @dramaticlly !  Left some more minor comments, mostly in comments as its a bit of a complex algorithm.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java:
```diff
@@ -0,0 +1,179 @@
+
+  private DeleteFile deleteFileWrapper(StructType sparkFileType, Row row) {
+    int specId = row.getInt(row.fieldIndex("spec_id"));
+    Types.StructType combinedFileType = DataFile.getType(Partitioning.partitionType(table));
```
Nit: is it combinedPartitionType?

dramaticlly replied:
I see where the confusion is from, this [`DataFile.getType(partitionType)`](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/DataFile.java#L108-L128) returns a struct type of a file with provided partition type. 

Here it's trying to project from (file with unioned partition) to (file with desired partition spec) for SparkDeleteFile

szehon-ho reviewed (approved):
This looks great to me, @dramaticlly one more comment

dramaticlly commented:
Thanks everyone for the review and input, special thanks to @aokolnychyi for optimized algorithm and @szehon-ho as original author and detailed review!
